### PR TITLE
fix: https://bugzilla.zimbra.com/show_bug.cgi?id=107013

### PR DESCRIPTION
**Pull request by ghost**
I have been ignoring bug 107013 for a while now, but in 8.7.7 type errors are shown to the end user (in addition to the browser console). Making this bug more annoying.

The bug is a result of using `disposeOnPopDown: true` when creating ZmDialog, after the dialog is disposed, and a resize event occurs, type errors are thrown. Because the dialog no longer exist.

To avoid the error I added `if(sizeThis)`, that will be `undefined` if the dialog is no longer existing.

### CODE REVIEW

**Comment by ghost** [comment]
@gordyt can you take care of the review?

**Comment by silentsakky** [comment]
added ui team for review

**Comment by silentsakky** [comment]
I think actual issue here is that after disposing the dialog we are still having reference to it, which is giving an error here, can we identify where it is not doing the proper cleanup?

**Comment by ghost** [comment]
Probably, but I would need some guidance to know where to begin.

**Comment by tarangkhandelwal** [comment]
It seems like one of the handler for the dialog is getting called when DwtShell is resized even after the Dialog is disposed. Looks like a case of improper cleaning of listeners while Disposing.

**Comment by kalpeshhpatel** [comment]
It seems to be core level issue.

When dialog opens, using `DwtBaseDialog.prototype.popup` method, it registers one resize event listeners  here: https://github.com/Zimbra/zm-ajax/blob/develop/WebRoot/js/ajax/dwt/widgets/DwtBaseDialog.js#L245

However, when dialog popdown, `DwtBaseDialog.prototype.popdown` method tries to remove previously registered resize event handlers, here: https://github.com/Zimbra/zm-ajax/blob/develop/WebRoot/js/ajax/dwt/widgets/DwtBaseDialog.js#L317. In this case, this event handlers, is not removing from set of all event handlers. And this is main issue.

So next time, when we resize browser window, and if element is `disposed` (`disposeOnPopDown: true `),  resize event handler of that element still executes, and it throws and error, as it tries to use element's size (`undefined` in this case).

(For element which are not `disposed`, it won't give any error in console, as the element is still present in DOM)

Here is the screenshot of call stack for reference, which maybe useful to others: 

image

As you can see in screenshot, at line number 411, we are comparing `member` which is event handler function to remove, with one of the listeners. And in this case, it is always coming out as `false`

@silentsakky @tarangkhandelwal @SatyamKumarOnline , let me know if you have any input.

Thanks

**Comment by kalpeshhpatel** [comment]
Current fix looks okay to me, as at-least it solves the issue https://bugzilla.zimbra.com/show_bug.cgi?id=107013 and has additional check which prevents trying to access undefined property. 

Actual issue is altogether different and should be tracked separately IMO.
 
@silentsakky @tarangkhandelwal @SatyamKumarOnline , let me know if you have any input.

Thanks

File: WebRoot/js/ajax/dwt/widgets/DwtControl.js [hunk 1]
```diff
@@ -2769,26 +2769,29 @@ function() {
...
-	if ((y + sizeThis.y) > sizeShell.y) {
-		y = sizeShell.y - sizeThis.y;
-	}
-	this.setLocation(x, y);
+      this._checkState();
+      var sizeShell = this.shell.getSize();
+      var sizeThis = this.getSize();
+      var x, y;
+      if(sizeThis)
+      {
```
Comment (kalpeshhpatel) [inline WebRoot/js/ajax/dwt/widgets/DwtControl.js]: @barrydegraaff , can you please make opening of `{` on the same line as `if`, like 
`if {`. 

This is to make sure, we all are on same coding convention. 

Thanks

**Comment by tarangkhandelwal** [comment]
Agree with @kalpeshhpatel  comment.
 DwtControl.prototype._position should not even be getting called if the dialog has been disposed.
The issue here seems to be with resize handler not getting unregistered on dispose of the Dialog ( DwtBaseDialog widget). 

I suggest to go with this fix: 
fix

**Comment by ghost** [comment]
Suggested fix from @tarangkhandelwal works well, I also left my original patch in the code, with the coding style suggested by @kalpeshhpatel.

File: WebRoot/js/ajax/dwt/widgets/DwtControl.js [hunk 2]
```diff
@@ -2769,26 +2769,28 @@ function() {
...
-	}
-	if ((y + sizeThis.y) > sizeShell.y) {
-		y = sizeShell.y - sizeThis.y;
-	}
-	this.setLocation(x, y);
+      this._checkState();
+      var sizeShell = this.shell.getSize();
+      var sizeThis = this.getSize();
+      var x, y;
+      if(sizeThis) {
```
Comment (tarangkhandelwal) [inline WebRoot/js/ajax/dwt/widgets/DwtControl.js]: space after if.

Should be:
` if (sizeThis) {`

**Comment by ghost** [comment]
@tarangkhandelwal sorry about that, here goes.

File: WebRoot/js/ajax/dwt/widgets/DwtBaseDialog.js [hunk 3]
```diff
@@ -242,7 +242,8 @@ function(loc) {
 	kbMgr.pushTabGroup(this._tabGroup);
 	kbMgr.pushDefaultHandler(this);
 
-	DwtShell.getShell().addListener(DwtEvent.CONTROL, this._resizeHdlr.bind(this));
+	this.shellResizeHandler = this._resizeHdlr.bind(this);
```
Comment (kalpeshhpatel) [inline WebRoot/js/ajax/dwt/widgets/DwtBaseDialog.js]: @barrydegraaff , can we move `this.shellResizeHandler = this._resizeHdlr.bind(this);` definition to object's constructor function.

Thanks

Comment (ghost) [reply WebRoot/js/ajax/dwt/widgets/DwtBaseDialog.js]: Maybe not 
TypeError: Cannot read property 'isListenerRegistered' of undefined at DwtShell._resizeHdlr 

if (shell.isListenerRegistered(DwtEvent.CONTROL)) {

At least its another puzzle.

**Comment by ghost** [comment]
@kalpeshhpatel can we leave it like this? So we make the next release?

**Comment by kalpeshhpatel** [comment]
@barrydegraaff , I just tried by moving definition for `shellResizeHandler ` to constructor and it worked for me. I am attaching patch for this change set.
Can you please see it it works for you?


[DwtBaseDialog.patch.zip](https://github.com/Zimbra/zm-ajax/files/947024/DwtBaseDialog.patch.zip)

Thanks

**Comment by ghost** [comment]
@kalpeshhpatel I am sorry, but that throws the same type errors as I mentioned [here](https://github.com/Zimbra/zm-ajax/pull/5#discussion_r111925922), in DwtShell._resizeHdlr at shell.isListenerRegistered that has something to do with where the listener is initialized.

**Comment by silentsakky** [comment]
@barrydegraaff I have updated your PR with working code, I have verified changes using pgp-zimlet and it is not throwing any JS errors after closing the dialogs, So I guess it should work fine, Please test from your side also

If you find any issues still then let us know the steps to reproduce

Below is the summary of changes that I did in your PR
- moved binding of this. _resizeHdlr to contructor so it will be reused everytime we create a new dialog
- squash 4 commits into 1, to make sure it doesn't pollute commit history of zm-ajax with unnecessary commits
- used tabs instead of spaces in function DwtControl._position

**Comment by ghost** [comment]
Thanks, I will create a clean PR next week, also please note that the pgp-zimlet includes an override to fix the current issue on the 8.7.7 release, so that is not the best way to test.
https://github.com/Zimbra-Community/pgp-zimlet/blob/stable/tk_barrydegraaff_zimbra_openpgp/tk_barrydegraaff_zimbra_openpgp.js#L874

**Comment by silentsakky** [comment]
I know and I have checked after removing that workaround

**Comment by ghost** [comment]
@silentsakky thanks, I verified the patch on a fresh build and it works well. Please proceed with the next review.

**Comment by silentsakky** [comment]
@barrydegraaff thanks for validating the fix, I have created a testing ticket for this fix, so whenever it is tested by QA then we will merge it to develop branch

**Review by silentsakky — APPROVED** [review]

**Comment by ghost** [comment]
@SatyamKumarOnline @silentsakky
Any news on when this will be merged, it has been a while now.

Also sometimes Zimbra releases public beta's, maybe this can be included.

Thanks, Barry

**Review by tarangkhandelwal — APPROVED** [review]
Looking good for testing and merging to the codebase.

**Comment by silentsakky** [comment]
@barrydegraaff Sorry for delay in merging this PR, we have created task to test this fix and merge it, most probably we will be taking it in next sprint

**Comment by silentsakky** [comment]
Fix had been merged to develop
https://github.com/Zimbra/zm-ajax/commit/a2119c20cd3919cbe769ddb760bd991f05351c7b

**Pull request closed without merging**